packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol

```diff
@@ -0,0 +1,119 @@
+pragma solidity 0.4.25;
+pragma experimental "ABIEncoderV2";
+
+import "../libs/LibStateChannelApp.sol";
+import "../libs/LibSignature.sol";
+import "../libs/LibStaticCall.sol";
+
+import "./MAppRegistryCore.sol";
+import "./MAppCaller.sol";
+
+
+contract MixinVirtualAppSetState is
+  LibSignature,
+  LibStateChannelApp,
+  MAppRegistryCore,
+  MAppCaller
+{
+
+  /// signatures[intermediaryIdx], instead of signing a message that authorizes
+  /// a state update with a given stateHash, signs a message that authorizes all
+  /// updates with nonce < nonceExpiry
+  struct VirtualAppSignedStateHashUpdate {
+    bytes32 stateHash;
+    uint256 nonce;
+    uint256 timeout;
+    bytes signatures;
+    uint256 intermediaryIdx;
+    uint256 nonceExpiry;
+  }
+
+  function setState(
+    AppIdentity appIdentity,
+    VirtualAppSignedStateHashUpdate req
+  )
+    public
+  {
+    bytes32 id = computeAppIdentityHash(appIdentity);
+
+    AppChallenge storage challenge = appStates[id];
+
+    require(
+      challenge.status == AppStatus.ON,
+      "setState was called on an app that is either in DISPUTE or OFF"
+    );
+
+    require(
+      correctKeysSignedTheStateUpdate(
+        id,
+        appIdentity.signingKeys,
+        req
+      ),
+      "Call to setState included incorrectly signed state update"
+    );
+
+    require(
+      req.nonce > challenge.nonce,
+      "Tried to call setState with an outdated nonce version"
+    );
+
+    require(
+      req.nonce < req.nonceExpiry,
+      "Tried to call setState with nonce greater than intemediary nonce expiry");
+
+    challenge.status = req.timeout > 0 ? AppStatus.DISPUTE : AppStatus.OFF;
+    challenge.appStateHash = req.stateHash;
+    challenge.nonce = req.nonce;
+    challenge.finalizesAt = block.number + req.timeout;
+    challenge.disputeNonce = 0;
+    challenge.disputeCounter += 1;
+    challenge.latestSubmitter = msg.sender;
+  }
+
+  function correctKeysSignedTheStateUpdate(
+    bytes32 id,
+    address[] signingKeys,
+    VirtualAppSignedStateHashUpdate req
+  )
+    private
+    pure
+    returns (bool)
+  {
+    bytes32 digest1 = computeStateHash(
+      id,
+      req.stateHash,
+      req.nonce,
+      req.timeout
+    );
+
+    bytes32 digest2 = keccak256(
+      abi.encodePacked(
+        byte(0x19),
+        id,
+        req.stateHash,
+        req.nonceExpiry,
+        req.timeout,
+        byte(1)
```

Comment:
Why?

Reply:
Otherwise signatures of  the two types cant be distinguished